### PR TITLE
Use packed structures in jemalloc by using the offset field

### PR DESCRIPTION
**Pull request by arichardson**

### CODE REVIEW

File: contrib/jemalloc/include/jemalloc/jemalloc_FreeBSD.h [hunk 1]
```diff
@@ -73,7 +73,7 @@
 #endif
 #ifdef __mips__
 #ifdef __mips_n64
-#  define LG_VADDR		64
+#  define LG_VADDR		48
```
Comment (nwf) [inline contrib/jemalloc/include/jemalloc/jemalloc_FreeBSD.h]: At least for the "old" pmap code, which is all we've ever used, this is in fact 39, I think.  48 is probably sane and I don't think lowering it to the "real" value buys us anything?

File: contrib/jemalloc/include/jemalloc/internal/rtree.h [hunk 2]
```diff
@@ -73,7 +71,9 @@ struct rtree_leaf_elm_s {
 	 *   00000000 xxxxxxxx eeeeeeee [...] eeeeeeee eeee000b
 	 */
```
Comment (nwf) [inline contrib/jemalloc/include/jemalloc/internal/rtree.h]: This comment probably wants some updating.  Something like...

```
On CHERI, where manipulation of the high bits is tricky, we are nevertheless able to
squeeze all metadata into a capability pointing to the extent.  We always mean to
point to the base (i.e., a zero offset relative to the start) of an extent, and so can
safely abuse the CHERI capability offset field to store at least 9 bits, even in the
CHERI Concentrate architectural encoding.  Conveniently, 9 bits are exactly what
we need.  The extent pointer (e) is recovered by setting the offset to 0, while the
slab flag (b) is the LSB of the offset and the size index (x) is the next 8 bits above
that; all further bits of the offset are maintained 0.
```

File: contrib/jemalloc/include/jemalloc/internal/rtree.h [hunk 3]
```diff
@@ -211,13 +216,23 @@ rtree_leaf_elm_bits_extent_get(uintptr_t bits) {
 }
 
 JEMALLOC_ALWAYS_INLINE szind_t
-rtree_leaf_elm_bits_szind_get(vaddr_t bits) {
+rtree_leaf_elm_bits_szind_get(uintptr_t bits) {
+#    ifdef __CHERI_PURE_CAPABILITY__
+	/* Lowest bit of offset is the boolean flag -> shift by one for szind */
+	return (szind_t)(cheri_getoffset((void*)bits) >> 1);
```
Comment (nwf) [inline contrib/jemalloc/include/jemalloc/internal/rtree.h]: In the original jemalloc encoding, the bits above szindex were always assumed to be 0.  Presumably that's true here, too, but either an explicit mask or a comment back-reffing the paragraph I suggested above might be a good idea.  If szind_t is uint8_t or such, the mask should be a no-op, right?

Comment (arichardson) [reply contrib/jemalloc/include/jemalloc/internal/rtree.h]: I've added an assertion

File: contrib/jemalloc/include/jemalloc/internal/rtree.h (hunk 2 above)
Comment (arichardson) [reply contrib/jemalloc/include/jemalloc/internal/rtree.h]: Thanks that looks good.

**Pull request merged**